src/core_plugins/kibana/public/dashboard/panel/panel_header/panel_actions/build_context_menu.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * Loops through allActions and extracts those that belong on the given contextMenuPanelId
+ * @param {string} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} allActions
+ */
+function getActionsForPanel(contextMenuPanelId, allActions) {
+  return allActions.filter(action => action.parentPanelId === contextMenuPanelId);
+}
+
+/**
+ * @param {String} contextMenuPanelId
+ * @param {Array.<DashboardPanelAction>} actions
+ * @param {Embeddable} embeddable
+ * @param {ContainerState} containerState
+ * @return {{
+ *   Array.<EuiContextMenuPanelItemShape> items - panel actions converted into the items expected to be on an
+ *     EuiContextMenuPanel,
+ *   Array.<EuiContextMenuPanelShape> childPanels - extracted child panels, if any actions also open a panel. They
+ *     need to be moved to the top level for EUI.
+ *  }}
+ */
+function buildEuiContextMenuPanelItemsAndChildPanels({ contextMenuPanelId, actions, embeddable, containerState }) {
+  const items = [];
+  const childPanels = [];
+  const actionsForPanel = getActionsForPanel(contextMenuPanelId, actions);
+  actionsForPanel.forEach(action => {
+    const isVisible = action.isVisible({ embeddable, containerState });
+    if (!isVisible) {
+      return;
+    }
+
+    let childPanelToOpenOnClick;
+    if (action.childContextMenuPanel) {
+      childPanels.push(
+        ...buildEuiContextMenuPanels({
+          contextMenuPanel: action.childContextMenuPanel,
+          actions,
+          embeddable,
+          containerState
+        }));
+      childPanelToOpenOnClick = action.childContextMenuPanel.id;
+    }
+
+    items.push(convertPanelActionToContextMenuItem(
+      {
+        action,
+        contextMenuPanelId: childPanelToOpenOnClick,
+        containerState,
+        embeddable
+      }));
+  });
+
+  return { items, childPanels };
+}
+
+/**
+ * Transforms a DashboardContextMenuPanel to the shape EuiContextMenuPanel expects, inserting any registered pluggable
+ * panel actions.
+ * @param {DashboardContextMenuPanel} contextMenuPanel
+ * @param {Array.<DashboardPanelAction>} actions to build the context menu with
+ * @param {Embeddable} embeddable
+ * @param {ContainerState} containerState
+ * @return {Object} An object that conforms to EuiContextMenuPanelShape in elastic/eui
+ */
+export function buildEuiContextMenuPanels(
+  {
+    contextMenuPanel,
+    actions,
+    embeddable,
+    containerState
+  }) {
+  const euiContextMenuPanel = {
+    id: contextMenuPanel.id,
+    title: contextMenuPanel.title,
+    items: [],
+    content: contextMenuPanel.getContent({ embeddable, containerState }),
+  };
+  let contextMenuPanels = [euiContextMenuPanel];
```

Comment:
This can become a const, as all we're doing is concatenating and returning, so instead:

```js
const contextMenuPanels = [euiContextMenuPanel];

// later
return contextMenuPanels.concat(childPanels);
```

(I have an irrational war against 'let' whenever I see it, so feel free to ignore.)